### PR TITLE
feat: Add percentile function to tensorlib

### LINKED ISSUE #815: Add a percentile function to tensorlib
# Description

First brought up in PR #790, there should be a percentile function added to the tensor backends that adheres to [NumPy's `percentile`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.percentile.html). For reference for the other backnds:

- [`tfp.stats.percentile`](https://www.tensorflow.org/probability/api_docs/python/tfp/stats/percentile)
- PyTorch doesn't have anything yet, but it is part of the NumPy functionality request rollup ([PyTorch Issue 38349](https://github.com/pytorch/pytorch/issues/38349)) and should be implemented in [PyTorch PR 39417](https://github.com/pytorch/pytorch/pull/39417).
- [`jax.numpy.percentile`](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.percentile.html)

Currently PyTorch doesn't support anything that matches these other `percentile` implimentations. The [`torch.distributions.distribution.icdf`](https://pytorch.org/docs/stable/distributions.html#torch.distributions.distribution.Distribution.icdf) gives the

> the inverse cumulative density/mass function evaluated at value

which is what is desired as [`np.percentile`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.percentile.html) 

> compute[s] the q-th percentile of the data along the specified axis

but the `icdf` is a distributions method, and not a tensor method.

### LINKED ISSUE #1573: Tracking of TensorFlow and Python 3.9 harmonization
This Issue to to track the line of PRs and Issues that need to be done in order to get TensorFlow harmonized with the other backends such that it can support Python 3.9. This will allow for development to continue while we decide how best to handle Issue #293.

- Issue #997 and PR #1001
- Issue #1515 and PR #1574
- Issue #815 and PR #817

### PR DESCRIPTION
**Pull request by matthewfeickert**
# Description

* ~~Partially addresses Issue #815~~
* Resolves #815
* Resolves #1573

Add a `percentile` function to the tensor backends that adheres to [NumPy's `numpy.percentile`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.percentile.html) API.

~~To get quotation marks to look okay for the interpolation method section of the docstrings I used `\\` to escape them given [this Stack Overflow question](https://stackoverflow.com/questions/17965655/how-to-escape-single-quotes-in-restructuredtext-when-converting-to-html-using-sp). There may be a better way.~~

Docs build at: https://pyhf--817.org.readthedocs.build/en/817/api.html#backends

# Checklist Before Requesting Reviewer

- [x] Tests are passing
- [x] "WIP" removed from the title of the pull request
- [x] Selected an Assignee for the PR to be responsible for the log summary

# Before Merging

For the PR Assignees:

- [x] Summarize commit messages into a comprehensive review of the PR

```
* Add percentile function to the tensor backends
* Add tests for percentile and its interpolation methods
   - JAX requires additional dtype support with the 'linear' interpolation method
     c.f. https://github.com/google/jax/issues/8513
   - PyTorch has yet to implement interpolation method options
   - c.f. https://github.com/scikit-hep/pyhf/issues/1693
```

### CODE REVIEW

**Comment by matthewfeickert** [comment]
Note to self: Should give more information in docstrings of interpolation methods and note what is the default interpolation method ("linear").

File: src/pyhf/tensor/pytorch_backend.py [hunk 1]
```diff
@@ -245,6 +245,45 @@ def log(self, tensor_in):
...
+                - \\'midpoint\\': Not yet implemented in PyTorch
+
+                - \\'nearest\\': Not yet implemented in PyTorch
+
+        Returns:
+            PyTorch tensor: The value of the :math:`q`-th percentile of the tensor along the specified axis.
+
+        """
+        # Interpolation options not yet supported
+        # c.f. https://github.com/pytorch/pytorch/issues/38349
```
Comment (kratsg) [inline src/pyhf/tensor/pytorch_backend.py:284]: should be out with pytorch/pytorch#48523 being done?

Comment (kratsg) [reply src/pyhf/tensor/pytorch_backend.py:284]: Nope -- need 1.7.2 (https://github.com/pytorch/pytorch/releases/tag/v1.7.1)

Comment (matthewfeickert) [reply src/pyhf/tensor/pytorch_backend.py:284]: Yeah, sadly. I just keep rebasing these PRs so that they don't fall too far behind and make it harder later on.

File: tests/test_tensor.py [hunk 2]
```diff
@@ -370,6 +370,61 @@ def test_boolean_mask(backend):
...
+    assert tb.tolist(tb.percentile(a, 50)) == 3.5
+    assert tb.tolist(tb.percentile(a, 100)) == 10
+    assert tb.tolist(tb.percentile(a, 50, axis=1)) == [7.0, 2.0]
+
+
+# FIXME: PyTorch doesn't yet support interpolation schemes other than "linear"
+# c.f. https://github.com/pytorch/pytorch/pull/59397
+@pytest.mark.skip_pytorch
+@pytest.mark.skip_pytorch64
+@pytest.mark.skip_jax
```
Comment (matthewfeickert) [inline tests/test_tensor.py:388]: Skipping PyTorch as interpolation kwargs aren't implemented yet as of `torch` `v1.10.0`. Skipping JAX so as to have separate test below given floating point issues with `"linear"` option.

File: tests/test_tensor.py [hunk 3]
```diff
@@ -370,6 +370,61 @@ def test_boolean_mask(backend):
...
+    a = tb.astensor([[10, 7, 4], [3, 2, 1]])
+
+    assert tb.tolist(tb.percentile(a, 50, interpolation="linear")) == 3.5
+    assert tb.tolist(tb.percentile(a, 50, interpolation="nearest")) == 3.0
+    assert tb.tolist(tb.percentile(a, 50, interpolation="lower")) == 3.0
+    assert tb.tolist(tb.percentile(a, 50, interpolation="midpoint")) == 3.5
+    assert tb.tolist(tb.percentile(a, 50, interpolation="higher")) == 4.0
+
+
+@pytest.mark.only_jax
```
Comment (matthewfeickert) [inline tests/test_tensor.py:400]: Hopefully these tests are tolerable at the moment, though obviously not ideal.

**Review by kratsg — APPROVED** [review]

**Comment by matthewfeickert** [comment]
RE: the deviations seen for JAX compared to all other backends, I opened up https://github.com/google/jax/issues/8513.

**Comment by matthewfeickert** [comment]
Issue #1693 exists to monitor the PyTorch and JAX issues, so this can PR can move forward.
